### PR TITLE
[SampleProfileLoader] Fix integer overflow in generateMDProfMetadata

### PR DESCRIPTION
**Pull request by omern1**
This patch fixes an integer overflow in the SampleProfileLoader pass.
The issue occurs when weights are saturated and Profi isn't being used.

This patch also adds a newline to a debug message to make it more
readable.

### CODE REVIEW

**Review by WenleiHe — APPROVED** [review]
lgtm, thanks.

**Comment by WenleiHe** [comment]
cc @MatzeB @wlei-llvm this one is narrow, but an evidence that we're not out of the woods yet for counts overflow.

**Comment by omern1** [comment]
@WenleiHe Do you know why the decision was made to use 32 bit values for metadata?

**Comment by WenleiHe** [comment]
> @WenleiHe Do you know why the decision was made to use 32 bit values for profile metadata?

I don't know the history (@david-xl may know), but I actually don't think using int32 is problematic by itself. With int64, of course it's easier, but it comes with size/memory cost. 

I think that int32 should be enough, because for profile counts, we don't necessarily care about the absolute values. What's important is 1) we preserve the relative difference in counts, and 2) we have good resolution on counts difference. We may need to normalize counts to overflow overflow/saturation, which we may not be doing a good job..

**Comment by david-xl** [comment]
> > @WenleiHe Do you know why the decision was made to use 32 bit values for profile metadata?
> 
> I don't know the history (@david-xl may know), but I actually don't think using int32 is problematic by itself. With int64, of course it's easier, but it comes with size/memory cost.
> 
> I think that int32 should be enough, because for profile counts, we don't necessarily care about the absolute values. What's important is 1) we preserve the relative difference in counts, and 2) we have good resolution on counts difference. We may need to normalize counts to overflow overflow/saturation, which we may not be doing a good job..

right -- 'weight' is not 'count'. It is scaled from edge count from the very beginning. The function entry's entry count on the other hand is 64 bit. Ideally the weight representation details should not be exposed and all updates be done via APIs.

File: llvm/lib/Transforms/IPO/SampleProfile.cpp
Comment (MatzeB) [inline llvm/lib/Transforms/IPO/SampleProfile.cpp:1723]: This will change the value of `Weight` below in line 1735 (or 1732 before). Probably should assign to some temporary variable.

File: llvm/test/Transforms/SampleProfile/overflow.ll
Comment (MatzeB) [inline llvm/test/Transforms/SampleProfile/overflow.ll:3]: Can you try using `utils/update_analyze_test_checks.py` and auto-generate the `CHECK:` lines that makes for more robust tests? We haven't mass-edited older tests for this style, but it should be used for newly added tests.

You can find an example in `test/Analysis/BranchProbabilityInfo/basic.ll`.

File: llvm/test/Transforms/SampleProfile/overflow.ll
Comment (MatzeB) [inline llvm/test/Transforms/SampleProfile/overflow.ll:7]: this function name doesn't match the IR... Though FWIW this seems so simple a function, can just as well remove the whole "Original Source" part in this case as far as I am concerned...

**Review by MatzeB — COMMENTED** [review]
Wow, good catch!

Surprising this wasn't noticed before, I guess loops that actually run for 4 billion iterations are rare outside of microbenchmarks...

**Review by MatzeB — CHANGES_REQUESTED** [review]
Adding "request changes" as this changing of the `Weight` variable instead of temporary seems dangerous.

File: llvm/lib/Transforms/IPO/SampleProfile.cpp
Comment (WenleiHe) [inline llvm/lib/Transforms/IPO/SampleProfile.cpp:1717]: Unrelated to this particular fix, but I'm also wondering if we would benefit from proper scaling/normalization, instead of just saturate uint32_t in the case of overflow. We can do some sanity check to see how much we actually hit such saturation.

**Comment by omern1** [comment]
The BOLT test failure on the pre-commit CI seems unrelated to this patch. It's reproducible even without the patch.

File: llvm/lib/Transforms/IPO/SampleProfile.cpp
Comment (spupyrev) [inline llvm/lib/Transforms/IPO/SampleProfile.cpp:1722]: I'm curious if this path is still beneficial for some use cases? Can we trust profi results and drop this hack altogether? @omern1

Comment (WenleiHe) [reply llvm/lib/Transforms/IPO/SampleProfile.cpp:1722]: you mean delete the non-profi inference path completely? some people might still want a fall back.

Comment (spupyrev) [reply llvm/lib/Transforms/IPO/SampleProfile.cpp:1722]: i'm curious to learn about such use cases; are there some known examples where fallback is preferred?

Comment (omern1) [reply llvm/lib/Transforms/IPO/SampleProfile.cpp:1722]: Hi @spupyrev, I can't speak for all users of the non-profi inference path but from our perspective that would require analysis to understand the impact of profi on our codebases.

I think this discussion can be had separately from this patch.

If you / @WenleiHe  / @MatzeB are happy with the current state of this patch can I please get an LGTM?

Comment (spupyrev) [reply llvm/lib/Transforms/IPO/SampleProfile.cpp:1722]: Of course, I didn't plan to delay the patch, will stamp it. 
(Though if you care about perf and the experiments with profi aren't too time-consuming, please try it out and share the results with us. Ultimately it would be great to keep only one codepath here, but the change should be done carefully)

**Review by spupyrev — APPROVED** [review]

**Comment by spupyrev** [comment]
l see one failed test currently: 

     | LLVM :: Transforms/SampleProfile/fnptr.ll

Is this unrelated?

**Review by MatzeB — APPROVED** [review]
LGTM

FWIW: I also think that having two code paths for prof/non-profi is not great, but that shouldn't block a simple bug fix.

**Comment by MatzeB** [comment]
(obviously should fix the test before landing)

File: llvm/lib/Transforms/IPO/SampleProfile.cpp
Comment (omern1) [reply llvm/lib/Transforms/IPO/SampleProfile.cpp:1717]: Hi @WenleiHe, we probably do need some form of scaling here. In one of the codebases we (Sony) have we're seeing losses due to weights being saturated which will be helped scaling the values.

I'll try and address this in a future patch.

**Comment by omern1** [comment]
> l see one failed test currently:
> 
>   | LLVM :: Transforms/SampleProfile/fnptr.ll
> 
> Is this unrelated?

My bad, now fixed.

**Pull request merged**
